### ISSUE TITLE
pmpro_after_change_membership_level hook only fires when a level is cancelled

### ISSUE THREAD

**Issue by Abdullah8006** [strangerstudios/paid-memberships-pro#1541]
**pmpro_after_change_membership_level** hook isn't working as expected

When a user changes his memebership level the hook **isn't** fired.

The hook is only fired when a user **cancels** his level.

Steps to reproduce
1. Log in as a user.
2. Go to the memebership details page.
3. Try to change your memebership level to a new level.

Following is my code that's calling the above mentioned hook

```
add_action( 'pmpro_after_change_membership_level', 'discord_level_change', 10, 3);

function discord_level_change($level_id, $user_id, $cancel_level) {
	//Do Some discord level work
}
```

**Comment by mircobabini** [comment]
I must disagree. That hook is triggered every time the `pmpro_changeMembershipLevel` function is called.

**Comment by dparker1005** [comment]
Hi @Abdullah8006,

As @mircobabini mentioned, that hook should be triggered for even membership level change. Is this still an issue that you are seeing on your site or has this been resolved? If this is still happening, can you please let us know if this still happens after disabling all PMPro Add Ons and custom code on your site?

**Comment by ChrisStudy** [comment]
I think i had same issue, when the membership level is cancel it fired, however when user paid the membership or upgrade their membership level it does not fired

**Comment by dparker1005** [comment]
Hi @ChrisStudy,

Can you please confirm that this happens with all add ons/custom code inactive besides for the function that you are hooking in on `pmpro_after_change_membership_level`? What is the code that you are hooking in on that action?

**Comment by ChrisStudy** [comment]
Here is the function i hooked to,
function pmpro_after_change_author_level_update_posts_published( $level_id, $user_id ) {
	//get the user roles
	$usermeta = get_userdata($user_id);
	$user_roles = $usermeta->roles;
	echo "User level is ".$level_id."</br>";
	echo "What is the user's role now"."</br>";
	var_dump($user_roles)."</br>";
// somecode to put user's post from draft to published. There is a similar function we hooked when member role change to put there post to draft which is working. So this is the beginning  of the function I assume it should be trigger when user's role change, but these code only display when user's cancel their membership not re-subscribe their membership

}
add_action( 'pmpro_after_change_membership_level', 'pmpro_after_change_author_level_update_posts_published', 10,3);

**Comment by ChrisStudy** [comment]
My problems is solved, it does get trigger but for some reason the message from the function don't get to display on payment confirmation page. I guest it's for security issue or some how. Buy anyway thanks for helping

**Comment by dparker1005** [comment]
Hi @ChrisStudy,

I am glad to hear that the `pmpro_after_change_membership_level` hook is in fact running during checkout. I believe that the issue with the text not showing is that, after the checkout itself is processed on the checkout page, PMPro then redirects the user to the confirmation page. If you are outputting content on `pmpro_after_change_membership_level`, that content will be lost during the redirect.

**Comment by ChrisStudy** [comment]
@dparker1005 that's why I couldn't get my text, by the way, in the future is there a safe way to display text regardless of redirection? so it will be easier for debugging in the future in general not only for this plugin?

**Comment by dparker1005** [comment]
The best way to save these error logs regardless of redirection is by using the `error_log()` function.

As the `pmpro_after_change_membership_level` hook now appears to be working as expected, I am going to close this issue.
